docs/RingBuffer.md

```diff
@@ -0,0 +1,414 @@
+      uint32_t wait_err = GetLastError();
+      if (wait_err == /* terminal error */) {
+        // … log error …
+        break;
+      }
+      producer_offset = ReadAcquire64(prod_offset);
+      have_data = producer_offset > consumer_offset; // It's possible we still have data.
+      if (!have_data) continue;
+    } else { // We got notified of new data.
+      have_data = true;
+    }
+  }
+  uint64_t remaining = producer_offset - consumer_offset;
+
+  if (remaining == 0) {
+    have_data = false; // Caught up to producer.
+    continue;
+  } else if (remaining < sizeof(rb_header_t)) {
+    // Bad record or consumer offset out of alignment.
+    // … log error …
+    break;
+  }
+
+  // Check header flags first, then read/skip data and update offset.
+  rb_header_t header = (rb_header_t)(&rb_data[consumer_offset % rb_size]);
+  if (header.locked) { // Next record not ready yet, wait.
+    have_data = false;
+    continue;
+    // Or we could spin/poll on ((rb_header_t*)(&rb_data[consumer_offset % rb_size]))->locked.
+  }
+  if (!header.discarded) {
+    const rb_record_t *record = *(const rb_record_t*)(&rb_data[consumer_offset % rb_size]);
+    // Read data from record->data[0 ... record->length-1].
+    // … business logic …
+  } // Else it was discarded, skip and continue.
+
+  // Update consumer offset (and pad record length to multiple of 8).
+  consumer_offset += sizeof(rb_header_t) + (record->length + 7 & ~7);
+  WriteNoFence64(cons_offset,consumer_offset);
+}
+
+Exit:
+```
+
+### Simplified polling ringbuf consumer
+
+This consumer uses the newly added helpers to consume the ring buffer.
+
+```c
+// Initialize wait handle for map.
+HANDLE wait_handle = ebpf_ring_buffer_get_wait_handle(map_fd);
+if (!wait_handle) {
+    // … log error …
+    goto Exit;
+}
+
+uint32_t wait_err = 0;
+
+// Consumer loop.
+for(;;) {
+  for(; !(err=rb__get_record(prod,cons,&record)); rb__next_record(prod,cons)) {
+    // Data is now in record->data[0 ... record->length-1].
+    // … Do record handling here …
+  }
+  // 3 cases for err:
+  // 1) Ringbuf empty - Wait on handle, or poll for !rb__empty(prod,cons).
+  // 2) Record locked - Wait on handle, or spin/poll on header lock bit.
+  // 3) Corrupt record or consumer offset - Break (could flush to continue reading from next good record).
+  if (err!=E_EMPTY && err!=E_LOCKED) {
+    // … log error …
+    break;
+  }
+  DWORD wait_status = WaitForSingleObject(wait_handle, INFINITE);
+
+  if (wait_status != WAIT_OBJECT_0) { // No notification
+    wait_err = GetLastError();
+    if (wait_err == /* terminal error */) {
+      // … log error …
+      break;
+    }
+  }
+}
+
+```
+
+### Polling ring buffer consumer (linux-style)
+
+```c
+// sample callback
+int ring_buffer_sample_fn(void *ctx, void *data, size_t size) {
+  // … business logic to handle record …
+}
+
+// consumer code
+struct ring_buffer_opts opts;
+opts.sz = sizeof(opts);
+opts.flags = RINGBUF_FLAG_NO_AUTO_CALLBACK; //no automatic callbacks
```

Comment:
One should not need to opt into Linux behavior, it should be the default.
We want to avoid gratuitous differences.